### PR TITLE
Split GIF support out into an integration library

### PR DESCRIPTION
**Pull request by sjudd**
## Description
Moves GIF support into an integration library, leveraging the new annotation processor and GlideExtensions.

### CODE REVIEW

**Comment by sjudd** [comment]
I need to uncomment and fix the tests here.

File: integration/gifdecoder/src/main/java/com/bumptech/glide/integration/gifdecoder/GifFrameLoader.java [hunk 1]
```diff
@@ -1,7 +1,4 @@
-package com.bumptech.glide.load.resource.gif;
-
-import static com.bumptech.glide.request.RequestOptions.diskCacheStrategyOf;
-import static com.bumptech.glide.request.RequestOptions.signatureOf;
+package com.bumptech.glide.integration.gifdecoder;
```
Comment (TWiStErRob) [inline integration/gifdecoder/src/main/java/com/bumptech/glide/integration/gifdecoder/GifFrameLoader.java]: RequestOptions used around here could be cached?

File: integration/gifdecoder/src/main/java/com/bumptech/glide/integration/gifdecoder/RequestManagerExtension.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package com.bumptech.glide.integration.gifdecoder;
+
+import android.graphics.drawable.Drawable;
+import com.bumptech.glide.TransitionOptions;
+import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
+import com.bumptech.glide.request.RequestOptions;
+
+/**
+ * Testing.
```
Comment (TWiStErRob) [inline integration/gifdecoder/src/main/java/com/bumptech/glide/integration/gifdecoder/RequestManagerExtension.java]: ?

File: integration/gifdecoder/src/test/java/com/bumptech/glide/integration/gifdecoder/ByteBufferGifDecoderTest.java [hunk 3]
```diff
@@ -17,14 +14,15 @@
...
-import com.bumptech.glide.tests.GlideShadowLooper;
+import com.bumptech.glide.testlib.GlideShadowLooper;
 import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.util.ArrayList;
 import java.util.List;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
+import org.mockito.Matchers;
```
Comment (TWiStErRob) [inline integration/gifdecoder/src/test/java/com/bumptech/glide/integration/gifdecoder/ByteBufferGifDecoderTest.java]: is this a new code style? It makes test code with arg matchers hard to read.

Comment (TWiStErRob) [inline integration/gifdecoder/src/test/java/com/bumptech/glide/integration/gifdecoder/ByteBufferGifDecoderTest.java]: I think `Assert.*`, `Mockito.*`, `Matchers.*`, `ArgumentMatchers.*` etc should be always static imported.

File: integration/gifdecoder/src/test/java/com/bumptech/glide/integration/gifdecoder/GifDrawableTransformationTest.java [hunk 4]
```diff
@@ -46,36 +40,49 @@ public void testSetsTransformationAsFrameTransformation() {
...
   public void testEquals() throws NoSuchAlgorithmException {
-    doAnswer(new Util.WriteDigest("first")).when(wrapped)
-        .updateDiskCacheKey(isA(MessageDigest.class));
-    KeyAssertions.assertSame(transformation, new GifDrawableTransformation(wrapped, bitmapPool));
-
-    @SuppressWarnings("unchecked") Transformation<Bitmap> other = mock(Transformation.class);
-    doAnswer(new Util.WriteDigest("other")).when(other)
-        .updateDiskCacheKey(isA(MessageDigest.class));
-    KeyAssertions.assertDifferent(transformation, new GifDrawableTransformation(other, bitmapPool));
+//    Mockito.doAnswer(new WriteDigest("first")).when(wrapped)
```
Comment (TWiStErRob) [inline integration/gifdecoder/src/test/java/com/bumptech/glide/integration/gifdecoder/GifDrawableTransformationTest.java]: forgot this?

File: integration/gifencoder/build.gradle [hunk 5]
```diff
@@ -9,6 +9,7 @@ dependencies {
     testCompile "org.mockito:mockito-all:${MOCKITO_VERSION}"
     testCompile "org.robolectric:robolectric:${ROBOLECTRIC_VERSION}"
     testCompile "com.android.support:support-v4:${SUPPORT_V4_VERSION}"
+    compile project(path: ':integration:gifdecoder')
```
Comment (TWiStErRob) [inline integration/gifencoder/build.gradle]: `path: ` seems unnecessary, not used in other places

File: library/src/main/java/com/bumptech/glide/RequestManager.java [hunk 6]
```diff
@@ -14,7 +14,7 @@
 import android.view.View;
 import com.bumptech.glide.load.engine.DiskCacheStrategy;
 import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
-import com.bumptech.glide.load.resource.gif.GifDrawable;
+//import com.bumptech.glide.load.resource.gif.GifDrawable;
```
Comment (TWiStErRob) [inline library/src/main/java/com/bumptech/glide/RequestManager.java]: forgot?

File: library/src/main/java/com/bumptech/glide/RequestManager.java [hunk 7]
```diff
@@ -46,7 +46,7 @@
  */
 public class RequestManager implements LifecycleListener {
   private static final RequestOptions DECODE_TYPE_BITMAP = decodeTypeOf(Bitmap.class).lock();
-  private static final RequestOptions DECODE_TYPE_GIF = decodeTypeOf(GifDrawable.class).lock();
+//  private static final RequestOptions DECODE_TYPE_GIF = decodeTypeOf(GifDrawable.class).lock();
```
Comment (TWiStErRob) [inline library/src/main/java/com/bumptech/glide/RequestManager.java]: forgot?

File: library/src/main/java/com/bumptech/glide/load/resource/drawable/DrawableResource.java [hunk 8]
```diff
@@ -39,8 +39,8 @@ public final T get() {
   public void initialize() {
     if (drawable instanceof BitmapDrawable) {
       ((BitmapDrawable) drawable).getBitmap().prepareToDraw();
-    } else if (drawable instanceof GifDrawable) {
-      ((GifDrawable) drawable).getFirstFrame().prepareToDraw();
+//    } else if (drawable instanceof GifDrawable) {
```
Comment (TWiStErRob) [inline library/src/main/java/com/bumptech/glide/load/resource/drawable/DrawableResource.java]: forgot?

if you need it how about introducing an interface (so there's only need for BitmapDrawable and that interface for instanceof)

File: library/src/main/java/com/bumptech/glide/request/BaseRequestOptions.java [hunk 9]
```diff
@@ -24,10 +24,10 @@
 import com.bumptech.glide.load.resource.bitmap.Downsampler;
 import com.bumptech.glide.load.resource.bitmap.FitCenter;
 import com.bumptech.glide.load.resource.bitmap.VideoBitmapDecoder;
-import com.bumptech.glide.load.resource.gif.ByteBufferGifDecoder;
-import com.bumptech.glide.load.resource.gif.GifDrawable;
-import com.bumptech.glide.load.resource.gif.GifDrawableTransformation;
-import com.bumptech.glide.load.resource.gif.StreamGifDecoder;
+//import com.bumptech.glide.load.resource.gif.ByteBufferGifDecoder;
```
Comment (TWiStErRob) [inline library/src/main/java/com/bumptech/glide/request/BaseRequestOptions.java]: forgot?

File: library/src/test/java/com/bumptech/glide/GlideTest.java [hunk 10]
```diff
@@ -402,29 +400,29 @@ public void testNonDefaultModelWithRegisteredFactoryDoesNotThrow() {
...
-    String fakeUri = "content://fake";
-    InputStream testGifData = openResource("test.gif");
-    mockUri(Uri.parse(fakeUri), testGifData);
-
-    requestManager.as(byte[].class).apply(decodeTypeOf(GifDrawable.class)).load(fakeUri)
-        .into(target);
-
-    verify(target).onResourceReady(isA(byte[].class), isA(Transition.class));
-  }
+//
```
Comment (TWiStErRob) [inline library/src/test/java/com/bumptech/glide/GlideTest.java]: forgot?

File: library/src/test/java/com/bumptech/glide/load/resource/bitmap/BitmapDrawableTransformationTest.java [hunk 11]
```diff
@@ -1,5 +1,6 @@
 package com.bumptech.glide.load.resource.bitmap;
 
+import static com.bumptech.glide.testlib.Util.cast;
```
Comment (TWiStErRob) [inline library/src/test/java/com/bumptech/glide/load/resource/bitmap/BitmapDrawableTransformationTest.java]: this is static imported, but Mockito not in other classes?

File: library/src/test/java/com/bumptech/glide/load/resource/drawable/DrawableResourceTest.java [hunk 12]
```diff
@@ -50,15 +47,15 @@ public void testDoesNotReturnOriginalDrawableOnGet() {
...
-    GifDrawable expected = mock(GifDrawable.class);
-    Drawable.ConstantState constantState = mock(Drawable.ConstantState.class);
-    when(constantState.newDrawable()).thenReturn(expected);
-    when(drawable.getConstantState()).thenReturn(constantState);
-
-    assertEquals(expected, resource.get());
-
-    verify(drawable).getConstantState();
-    verify(constantState).newDrawable();
+//    GifDrawable expected = mock(GifDrawable.class);
```
Comment (TWiStErRob) [inline library/src/test/java/com/bumptech/glide/load/resource/drawable/DrawableResourceTest.java]: forgot?

File: testlib/src/main/java/com/bumptech/glide/testlib/Util.java [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+import org.robolectric.util.ReflectionHelpers;
+
+// FIXME move to testutil module
+/**
+ * Generic test utilities.
+ */
+public class Util {
```
Comment (TWiStErRob) [inline testlib/src/main/java/com/bumptech/glide/testlib/Util.java]: missing utility class no-constructor

File: testlib/src/main/java/com/bumptech/glide/testlib/Util.java [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.security.MessageDigest;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+import org.robolectric.util.ReflectionHelpers;
+
+// FIXME move to testutil module
```
Comment (TWiStErRob) [inline testlib/src/main/java/com/bumptech/glide/testlib/Util.java]: I think you just did this.

**Comment by sjudd** [comment]
Yeah the tests here need to be cleaned up (see my earlier comment).

GlideApp is part of the generated API. The generated API is currently available only for Applications (it's only generated when we find a RootGlideModule).  I can't just use com.bumptech.glide.Glide for the generated API because it would break all non-application projects. Even if the generated API were available for libraries, you'd have a name collision if both the library and the application generated the same class in the same package (ie com.bumptech.glide.Glide). 

The current thought on the generated API is:
1. Applications use the generated API by including a RootGlideModule and the name is different for the generated API to avoid confusion (defaults to GlideApp, but applications can change the name. For example the Google Photos version is called PhotosGlide).
2. Libraries (and Applications that don't want to deal with the complexity) use the existing Glide class and the fixed API.

Other solutions include:
1. Generating a Glide class for every library and/or application but base the Glide class on the package name (so you'd have com.google.android.apps.photos.Glide, com.bumptech.glide.integration.okhttp3.Glide etc). 
2. Generate com.bumptech.glide.Glide for applications and force libraries to use a renamed GenericGlide/BaseGlide etc.
3. Generate com.bumptech.glide.Glide for applications and generate some other named class for libraries.

The two basic limitations are:
1. We need an annotated class to generate an API and especially to generate an API in a custom package (ie in the library or application's package). Right now, the best I can think of is to generate APIs for applications/libraries with Root or Child GlideModules.
2. We need a coherent way for libraries and applications to migrate. We could take a hard line approach, force everyone to use at least one GlideModule (Root or Child) and only expose Glide and similar APIs via generated code. However, doing so would break a bunch of snapshot users (and Google apps), so we'd probably need to fork and abandon the current master as a branch to accomplish this.

Thoughts definitely welcome. The main reason I put this up so early was to hopefully get comments on the generated API and how it can work with libraries, especially those that extend the API.

**Comment by TWiStErRob** [comment]
Wow, nice writeup. I'll take a look on generated code, it's kind of unrelated to this issue. I think I still have your apt PR in my queue :)

**Comment by sjudd** [comment]
I think all the comments here are addressed.

**Pull request closed without merging**